src/ensembl/src/content/app/browser/track-panel/trackPanelActions.ts

```diff
@@ -1,20 +1,47 @@
-import { createAction, createStandardAction } from 'typesafe-actions';
+import { createAction } from 'typesafe-actions';
 import { ThunkAction } from 'redux-thunk';
 import { Action, ActionCreator } from 'redux';
+import { batch } from 'react-redux';
 
 import { RootState } from 'src/store';
 import { TrackType } from './trackPanelConfig';
 import browserStorageService from '../browser-storage-service';
 import { getBrowserActiveGenomeId } from '../browserSelectors';
 
-export const toggleTrackPanel = createStandardAction(
-  'track-panel/toggle-track-panel'
-)<boolean | undefined>();
+export const toggleTrackPanelForGenome = createAction(
+  'track-panel/toggle-track-panel',
+  (resolve) => {
+    return (isTrackPanelOpenedForGenome: {
+      activeGenomeId: string;
+      isTrackPanelOpened: boolean;
+    }) => resolve(isTrackPanelOpenedForGenome);
+  }
+);
+
+export const toggleTrackPanel: ActionCreator<
+  ThunkAction<void, any, null, Action<string>>
+> = (isTrackPanelOpened: boolean) => (dispatch, getState: () => RootState) => {
+  const activeGenomeId = getBrowserActiveGenomeId(getState());
+
+  if (!activeGenomeId) {
+    return;
+  }
+
+  dispatch(
+    toggleTrackPanelForGenome({
+      activeGenomeId,
+      isTrackPanelOpened
+    })
+  );
+};
```

Comment:
Just a thought. If you had a `getActiveTrackPanel` selector (my comment for selectors file), you could do something like this:

```
  const state = getState();
  const activeGenomeId = getBrowserActiveGenomeId(state);
  const activeTrackPanel = getActiveTrackPanel(state);

  // after check for nulls

  const payload = {
    [activeGenomeId]: {
      ...activeTrackPanel,
      isTrackPanelOpened
    }
   }

  dispatch(updateTrackPanels(payload));
```

I am not sure whether this approach will be better, but It seems that it will reduce the amount of code in actions and reducer.

Reply:
But won't this not tell us what part of the track panel state has changed in Redux (or analytics)?

Reply:
I have made changes to be somewhat similar to this.